src/clusterfuzz/_internal/system/archive.py

```diff
@@ -407,6 +412,103 @@ class ArchiveType:
   TAR_LZMA = 3
 
 
+@dataclasses.dataclass
+class CacheBlock:
+  """Represents a cache entry for the HttpZipFile."""
+  start: int
+  end: int
+  content: Sequence[bytes]
+
+
+class HttpZipFile:
+  """This class is a very simple file-object representation of a file over HTTP.
+  It uses the 'Accept-Ranges' feature of HTTP to fetch parts (or all) of the
+  file.
+  """
+
+  @staticmethod
+  def is_uri_compatible(uri: str) -> bool:
+    try:
+      res = requests.head(uri, timeout=5)
+      return res.headers.get('Accept-Ranges') is not None
+    except:
+      return False
+
+  def __init__(self, uri):
+    self.uri = uri
+    self.session = requests.Session()
+    # This slows down the overall performances.
+    self.session.headers.pop('Accept-Encoding')
+    resp = self.session.head(self.uri)
+    self.file_size = int(resp.headers.get('Content-Length', default=0))
+    self._current_block = CacheBlock(0, 0, [])
+    self._pos = 0
+    assert resp.headers.get('Accept-Ranges') is not None
+
+  def seekable(self) -> bool:
+    """Whether this is seekable.
+    """
+    return True
+
+  def seek(self, offset: int, from_what: int = 0):
+    """Provides a seek implementation.
+
+    Args:
+        offset: the offset
+        from_what: from where the offset should be computed. Defaults to 0.
+    """
+    if from_what == 0:
+      self._pos = offset
+    elif from_what == 1:
+      self._pos = self._pos + offset
+    else:
+      self._pos = self.file_size + offset
+    if self._pos > self.file_size:
+      self._pos = self.file_size
+    self._pos = max(self._pos, 0)
+
+  def tell(self) -> int:
+    """Provides a tell implementation. Returns the current curso position.
+
+    Returns:
+        the current cursor position.
+    """
+    return self._pos
+
+  def _fetch_from_http(self, start, end) -> Sequence[bytes]:
+    resp = self.session.get(self.uri, headers={'Range': f'bytes={start}-{end}'})
+    return resp.content
+
+  def _fetch_from_cache(self, start, end) -> Sequence[bytes]:
+    if self._current_block.start > start or self._current_block.end < end:
+      read_ahead_end = min(self.file_size - 1, end + REMOTE_HTTP_MIN_READ_SIZE)
```

Comment:
`REMOTE_HTTP_MIN_READ_SIZE` does not behave as I expected: I thought this would be the minimum size we would fetch in one request, but this is actually a fixed readahead value.

I thought:

* read [0, 1M) -> fetch [0, 50M)
* read [0, 100M) -> fetch [0, 100M)

But this does:

* read [0, 1M) -> fetch [0, 51M)
* read [0, 100M) -> fetch [0, 150M)

Which is a totally fair policy, let's just make it clearer to the reader. How about:

* rename `REMOTE_HTTP_MIN_READ_SIZE` -> `HTTP_READAHEAD_BYTES`
* document it with a comment explaining its behavior